### PR TITLE
earning oracle

### PR DESCRIPTION
**Pull request by samparsky**
An on-chain aggregator of paid out per channel for AdEx Network earners.

### CODE REVIEW

File: contracts/EarningOracle.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    constructor (IAdExCore _core) public {
+        core = _core;
+    }
+
+    function bulkUpdate(bytes32[] calldata channelIds, address[] calldata earners) external {
+        for(uint i = 0; i < channelIds.length; i++) {
+            bytes32 channelId = channelIds[i];
+            address earner = earners[i];
+            
+            require(tallied[earner][channelId] == false, 'ALREADY_TALLIED');
```
Comment (IvoPaunov) [inline contracts/EarningOracle.sol:25]: Consider using `if` instead `require`. With `require` you have to keep strict state of the updated channels. For example - there are 10 channels on the first bulk update, after some time there are 5 new channels and it is time to make new `bulkUpdate` but you don't keep record of the previous updated channels and make update with all 15 channels - and the update will be "reverted".

**Review by IvoPaunov — COMMENTED** [review]
Looks good, just I'm not sure for using `require` at `bulkUpdate` function as in the comment. It is good to call functions with checked and validated data and the updated channels can be retrieved from `tallied` mapping as we have channels and earners before the `bulkUpdate` and on the second thought using `require` is OK.

File: contracts/EarningOracle.sol (hunk 1 above)
Comment (samparsky) [reply contracts/EarningOracle.sol:25]: The record of previously updated channels will be kept off-chain.

Comment (IvoPaunov) [reply contracts/EarningOracle.sol:25]: Ok

File: contracts/mock/Token.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: agpl-3.0
```
Comment (Ivshti) [inline contracts/mock/Token.sol:1]: can't you use one of the other mock tokens, such as `tests/mock/Token`?

**Review by Ivshti — CHANGES_REQUESTED** [review]
Looks good except the token mock is not needed and it still needs a JS script that updates the oracle

**Comment by Ivshti** [comment]
shelving this cause in v5, withdrawnPerUser for a single channel will be sufficient for this

**Pull request closed without merging**